Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50915 Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}**
* #50744 Turn on batched grad testing for CriterionTest

Fixes #50584
Add a vectorize flag to torch.autograd.functional.jacobian and
torch.autograd.functional.hessian (default: False). Under the hood, the
vectorize flag uses vmap as the backend to compute the jacobian and
hessian, respectively, providing speedups to users.

Test Plan:
- I updated all of the jacobian and hessian tests to also use
vectorized=True
- I added some simple sanity check tests that check e.g. jacobian with
vectorized=False vs
jacobian with vectorized=True.
- The mechanism for vectorized=True goes through batched gradient
computation. We have separate tests for those (see other PRs in this
stack).

Differential Revision: [D26057674](https://our.internmc.facebook.com/intern/diff/D26057674)

zou3519 committed:
Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}

Fixes #50584
Add a vectorize flag to torch.autograd.functional.jacobian and
torch.autograd.functional.hessian (default: False). Under the hood, the
vectorize flag uses vmap as the backend to compute the jacobian and
hessian, respectively, providing speedups to users.

Test Plan:
- I updated all of the jacobian and hessian tests to also use
vectorized=True
- I added some simple sanity check tests that check e.g. jacobian with
vectorized=False vs
jacobian with vectorized=True.
- The mechanism for vectorized=True goes through batched gradient
computation. We have separate tests for those (see other PRs in this
stack).

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -5812,7 +5836,62 @@ def foo(x, y):
+        self._check_jacobian_vectorize_correctness(f, (x, y))
+
+    def _check_hessian_vectorize_correctness(self, f, inputs):
+        expected = autogradF.jacobian(f, inputs)
```
These should be `hessian`?

zou3519 replied:
Oops copy-paste error. Thanks for the catch!

albanD commented on test/test_autograd.py:
```diff
@@ -5827,30 +5906,36 @@ def bar3(a):
         inp = torch.rand(4)
         with self.assertRaisesRegex(TypeError, "The inputs given to hessian must be either a Tensor"):
-            res = autogradF.hessian(foo, (inp, 2))
+            res = autogradF.hessian(foo, (inp, 2), vectorize=True)
```
Should these be `vectorize=vectorize` ?

albanD commented on test/test_autograd.py:
```diff
@@ -5864,43 +5949,55 @@ def bar2(a):
         with self.assertRaisesRegex(RuntimeError, "Output 0 of the user-provided function does not require gradients."):
-            res = autogradF.hessian(foo, inp, strict=True)
-        res = autogradF.hessian(foo, inp, strict=False)
+            res = autogradF.hessian(foo, inp, strict=True, vectorize=True)
```
same `vectorize=vectorize`

albanD commented on test/test_autograd.py:
```diff
@@ -5918,23 +6015,23 @@ def bad_reducer(x):
 
         inputs = torch.rand(2, 2, requires_grad=True)
-        res = autogradF.hessian(pow_reducer, inputs, create_graph=True)
+        res = autogradF.hessian(pow_reducer, inputs, create_graph=True, vectorize=True)
```
same `vectorize=vectorize`?

albanD replied:
no worries!

albanD commented on torch/autograd/functional.py:
```diff
@@ -429,6 +436,42 @@ def jacobian(func, inputs, create_graph=False, strict=False):
     jacobian: Tuple[torch.Tensor, ...] = tuple()
+
+    if vectorize:
+        for outidx, out in enumerate(outputs):
```
contrary to what is said in the docstring, you actually call autograd.grad once per output here.
Is that expected?

albanD replied:
Can we actually do that in a single call by giving a tuple of output/grad_output?

zou3519 replied:
> contrary to what is said in the docstring, you actually call autograd.grad once per output here.
Is that expected?

For the docstring, I meant to say "... calls autograd.grad once per output".

> Can we actually do that in a single call by giving a tuple of output/grad_output?

I never thought about this before. It turns out that we can! (and probably should). Let me think about it for a bit and then update the PR

zou3519 replied:
So the answer is yes, we can and should just do one call. However, this comes with the following caveat:
- `vectorize=True` and `strict=True` doesn't quite work. This is because instead of returning `None` for autograd.grad, `vectorize=True` might cause it to return a tensor full of zeros.
- I am thinking of just raising an error message if the above two are passed in together.

albanD replied:
I guess an error is ok for now.
But in the future, I think it would be ok to get just 0s for these and not raise the errror?

zou3519 replied:
> But in the future, I think it would be ok to get just 0s for these and not raise the errror?

Is it OK for `strict=False` to cause different behavior when `vectorize` is True and when it is False? I thought that would have been strange

albanD replied:
You mean `strict=True` right?
Yes I agree with you that it is surprising and forbidding it is definitely the best move for now.

zou3519 replied:
Yes, I meant `strict=True`, my bad. That sounds good (y)

albanD commented on torch/autograd/functional.py:
```diff
@@ -429,6 +436,42 @@ def jacobian(func, inputs, create_graph=False, strict=False):
+                        vj[el_idx] = torch.zeros_like(inputs[el_idx])
+                return tuple(vj)
+
+            basis_vectors = torch.eye(flat_out.numel(), dtype=flat_out.dtype, device=flat_out.device)
```
nit: we don't have a `eye_like` function?

zou3519 replied:
Unfortunately we don't

albanD commented on torch/autograd/functional.py:
```diff
@@ -545,7 +595,7 @@ def jac_func(*inp):
         return jac
 
-    res = jacobian(jac_func, inputs, create_graph=create_graph, strict=strict)
+    res = jacobian(jac_func, inputs, create_graph=create_graph, strict=strict, vectorize=vectorize)
```
Should we also `vectorize` the call to jacobian in `jac_func` ?
Or that is not supported by vmap?

zou3519 replied:
The function that we are computing a hessian over only has 1 output with 1 element. Therefore the `jacobian` call in `jac_func` is equivalent to a `vjp` call and doesn't benefit from `vectorize=True`.

albanD replied:
Ho right, I forgot that we have this constraint! Sounds good!

zou3519 committed:
Update on "Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}"


Fixes #50584
Add a vectorize flag to torch.autograd.functional.jacobian and
torch.autograd.functional.hessian (default: False). Under the hood, the
vectorize flag uses vmap as the backend to compute the jacobian and
hessian, respectively, providing speedups to users.

Test Plan:
- I updated all of the jacobian and hessian tests to also use
vectorized=True
- I added some simple sanity check tests that check e.g. jacobian with
vectorized=False vs
jacobian with vectorized=True.
- The mechanism for vectorized=True goes through batched gradient
computation. We have separate tests for those (see other PRs in this
stack).

[ghstack-poisoned]

zou3519 commented on torch/autograd/functional.py:
```diff
@@ -364,7 +365,37 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
+    # in torch.autograd.functional.jacobian.
+    assert len(tensors) == len(tensor_numels)
+    assert len(tensors) > 0
+    total_numel = sum(tensor_numels)
+    diag_start_indices = (0, *torch.tensor(tensor_numels).cumsum(dim=0)[:-1].neg().unbind())
+    chunks = tuple(tensor.new_zeros(total_numel, tensor_numel)
+                   for tensor, tensor_numel in zip(tensors, tensor_numels))
+    for chunk, diag_start_idx in zip(chunks, diag_start_indices):
+        chunk.diagonal(diag_start_idx).fill_(1)
+    return chunks
```
I'm not sure if there's a more elegant way of doing this. We could construct an NxN identity matrix and then split it into chunks and finally coerce them to the correct device and dtype, but that won't perform well if `tensors` are on different devices or have different dtypes because it would perform copies

albanD replied:
I think this version is good.
And the comment to explain it is quite nice!
A single Tensor would increase the memory overhead quite a bit, especially when multiple devices/dtypes are involved.

albanD commented on torch/autograd/functional.py:
```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
+            for el_idx, vj_el in enumerate(vj):
+                if vj_el is not None:
+                    continue
+                vj[el_idx] = torch.zeros_like(inputs[el_idx])
```
nit: Shall we use `new_zeros` here as well?

zou3519 replied:
Is the suggestion `vj[el_idx] = vj[el_idx].new_zeros(inputs[el_idx].shape)`?

albanD replied:
`vj[el_idx] = inputs[el_idx].new_zeros(inputs[el_idx].shape)` actually as `vj[el_idx]` is None here. But yes.
I started trying to use the `new_*` variants as much as possible to avoid future errors with vmap. But maybe that's not really necessary?

zou3519 replied:
Oh, I see. Thanks for the clarification, I wasn't thinking being able to vmap through this.

`inputs[el_idx].new_zeros(inputs[el_idx].shape)` and `torch.zeros_like(inputs[el_idx])` are the same with respect to vmap: if `inputs[el_idx]` is a BatchedTensor, then that information gets propagated by `new_zeros` and `zeros_like`. What doesn't propagate information is `torch.zeros` because `torch.zeros` doesn't accept a tensor.

albanD replied:
Thanks for the clarification!

albanD commented on torch/autograd/functional.py:
```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
+        # we split each Tensor by output.
+        jacobian = tuple(
+            tuple(jac_out.view(output.shape + inp.shape)
+                  for jac_out, output in zip(jac.split(output_numels, dim=0), outputs))
```
nit: I would find this more readable if it was not a one liner to be honest :D

zou3519 replied:
It is more verbose if this were to be for loops, but yeah it should be more readable. I'll change it

albanD commented on torch/autograd/functional.py:
```diff
@@ -429,6 +467,85 @@ def jacobian(func, inputs, create_graph=False, strict=False):
+        # Step 3: The returned jacobian is one big tensor per input. In this step,
+        # we split each Tensor by output.
+        jacobian = tuple(
+            tuple(jac_out.view(output.shape + inp.shape)
```
shouldn't the view be `view(output.shape, inp.shape)`?

zou3519 replied:
There are two ways to use view:
- pass var-args ints: tensor.view(1, 2, 3, 4, 5)
- pass a list/tuple of ints: tensor.view([1, 2, 3, 4, 5])

`jac_out.view(output.shape + input.shape)` is the latter, but if we wanted to do the former we could do `view(*output.shape, *inp.shape)`

albanD replied:
Ho these are tuples, not numbers. My bad!

albanD commented on torch/autograd/functional.py:
```diff
@@ -364,7 +365,37 @@ def jvp(func, inputs, v=None, create_graph=False, strict=False):
+    assert len(tensors) > 0
+    total_numel = sum(tensor_numels)
+    diag_start_indices = (0, *torch.tensor(tensor_numels).cumsum(dim=0)[:-1].neg().unbind())
+    chunks = tuple(tensor.new_zeros(total_numel, tensor_numel)
```
This made me realize that I think we're missing a small comment in the return type docstring: the Jacobian matrix will have the same dtype and device as its corresponding input.

zou3519 replied:
I can add one

zou3519 replied:
Docstring added to jacobian and hessian docs

zou3519 committed:
Update on "Add `vectorize` flag to torch.autograd.functional.{jacobian, hessian}"


Fixes #50584
Add a vectorize flag to torch.autograd.functional.jacobian and
torch.autograd.functional.hessian (default: False). Under the hood, the
vectorize flag uses vmap as the backend to compute the jacobian and
hessian, respectively, providing speedups to users.

Test Plan:
- I updated all of the jacobian and hessian tests to also use
vectorized=True
- I added some simple sanity check tests that check e.g. jacobian with
vectorized=False vs
jacobian with vectorized=True.
- The mechanism for vectorized=True goes through batched gradient
computation. We have separate tests for those (see other PRs in this
stack).

Differential Revision: [D26057674](https://our.internmc.facebook.com/intern/diff/D26057674)

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update!
